lib/utils/fs_test.go

```diff
@@ -26,6 +26,80 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestOpenFileSymlinks(t *testing.T) {
+	// symlink structure setup, this will produce the following structure under the temp directory created below:
+	// dir
+	// dir-s -> dir
+	// dir-s-s -> dir-s
+	// dir/file
+	// dir/file-s -> dir/file
+	// circular-s -> circular-s
+	// broken-s -> nonexistent
+	rootDir := t.TempDir()
+
+	dirPath := filepath.Join(rootDir, "dir")
+	err := os.Mkdir(dirPath, 0755)
+	require.NoError(t, err)
+
+	filePath := filepath.Join(dirPath, "file")
+	_, err = os.Create(filePath)
+	require.NoError(t, err)
+
+	symlinkToFile := filepath.Join(dirPath, "file-s")
+	err = os.Symlink(filePath, symlinkToFile)
+	require.NoError(t, err)
+
+	symlinkDir := filepath.Join(rootDir, "dir-s")
+	err = os.Symlink(dirPath, symlinkDir)
+	require.NoError(t, err)
+
+	symlinkToSymlinkDir := filepath.Join(rootDir, "dir-s-s")
+	err = os.Symlink(symlinkDir, symlinkToSymlinkDir)
+	require.NoError(t, err)
+
+	circularSymlink := filepath.Join(rootDir, "circular-s")
+	err = os.Symlink(circularSymlink, circularSymlink)
+	require.NoError(t, err)
+
+	brokenSymlink := filepath.Join(rootDir, "broken-s")
+	err = os.Symlink(filepath.Join(rootDir, "nonexistent"), brokenSymlink)
+	require.NoError(t, err)
+
+	// Tests below can not be done with t.Parallel due to the need for directory cleanup in defer
+	t.Run("directFileOpenAllowed", func(t *testing.T) {
+		filePath, err = filepath.EvalSymlinks(filePath)
+		require.NoError(t, err)
+		f, err := openFile(filePath, false)
```

Comment:
Please test the public functions instead.

Reply:
Now that the public functions include OS specific logic, I don't believe that would be a good fit for testing.

Reply:
If that's a concern, then we should be writing per-OS testing. I would still encourage you to test the public API and not internals.

Reply:
As in have the testing define the expected state depending on the OS it's run on?  Or have individual tests for each OS?

The first seems pretty reasonable if you would like me to update to that.  I will probably add it as additional testing rather than replace these.

Reply:
I meant the latter - we have a vanilla test suite for the "regular" scenarios, but test macOS vs Linux behavior separately. We have a few packages that behave like that, for example webauthn tends to vary between darwin, linux and windows.

Reply:
I am not sure I fully agree.  Because of the need to have a more complex setup that interacts with the filesystem I think it's best if this test is as centrally located as possible.

That said I did add testing around the public functions and the OS specific behaviors with hard links.  Let me know what you think.  Thank you!